File: benchmarkDS/CoT/CLAUDE_claude-sonnet-4-5-20250929/Lse58.cpp

#include <iostream>
#include <string>
#include <openssl/sha.h>
#include <openssl/rand.h>
#include <sstream>
#include <iomanip>
#include <regex>
#include <cstring>

class User {
private:
    std::string email;
    std::string passwordHash;
    std::string salt;
    bool isLoggedIn;

    std::string generateSalt() {
        unsigned char saltBytes[16];
        RAND_bytes(saltBytes, 16);
        
        std::stringstream ss;
        for(int i = 0; i < 16; i++) {
            ss << std::hex << std::setw(2) << std::setfill('0') << (int)saltBytes[i];
        }
        return ss.str();
    }

    std::string hashPassword(const std::string& password, const std::string& salt) {
        std::string combined = salt + password;
        unsigned char hash[SHA256_DIGEST_LENGTH];
        SHA256((unsigned char*)combined.c_str(), combined.length(), hash);
        
        std::stringstream ss;
        for(int i = 0; i < SHA256_DIGEST_LENGTH; i++) {
            ss << std::hex << std::setw(2) << std::setfill('0') << (int)hash[i];
        }
        return ss.str();
    }

    bool isValidEmail(const std::string& email) {
        std::regex pattern("^[A-Za-z0-9+_.-]+@[A-Za-z0-9.-]+\\\\.[A-Za-z]{2,}$");
        return std::regex_match(email, pattern);
    }

public:
    User(const std::string& email, const std::string& password) {
        this->email = email;
        this->salt = generateSalt();
        this->passwordHash = hashPassword(password, this->salt);
        this->isLoggedIn = false;
    }

    bool login(const std::string& email, const std::string& password) {
        if (this->email == email && this->passwordHash == hashPassword(password, this->salt)) {
            this->isLoggedIn = true;
            return true;
        }
        return false;
    }

    void logout() {
        this->isLoggedIn = false;
    }

    bool changeEmail(const std::string& oldEmail, const std::string& password, const std::string& newEmail) {
        if (!isLoggedIn) {
            std::cout << "Error: User must be logged in to change email" << std::endl;
            return false;
        }

        if (newEmail.empty() || !isValidEmail(newEmail)) {
            std::cout << "Error: Invalid new email format" << std::endl;
            return false;
        }

        if (this->email != oldEmail) {
            std::cout << "Error: Old email does not match" << std::endl;
            return false;
        }

        if (this->passwordHash != hashPassword(password, this->salt)) {
            std::cout << "Error: Incorrect password" << std::endl;
            return false;
        }

        this->email = newEmail;
        std::cout << "Email successfully changed to: " << newEmail << std::endl;
        return true;
    }

    std::string getEmail() const {
        return email;
    }

    bool isUserLoggedIn() const {
        return isLoggedIn;
    }
};

int main() {
    std::cout << "=== Test Case 1: Successful email change ===" << std::endl;
    User user1("user1@example.com", "password123");
    user1.login("user1@example.com", "password123");
    user1.changeEmail("user1@example.com", "password123", "newemail1@example.com");
    std::cout << "Current email: " << user1.getEmail() << "\\n" << std::endl;

    std::cout << "=== Test Case 2: Wrong old email ===" << std::endl;
    User user2("user2@example.com", "password456");
    user2.login("user2@example.com", "password456");
    user2.changeEmail("wrong@example.com", "password456", "newemail2@example.com");
    std::cout << "Current email: " << user2.getEmail() << "\\n" << std::endl;

    std::cout << "=== Test Case 3: Wrong password ===" << std::endl;
    User user3("user3@example.com", "password789");
    user3.login("user3@example.com", "password789");
    user3.changeEmail("user3@example.com", "wrongpassword", "newemail3@example.com");
    std::cout << "Current email: " << user3.getEmail() << "\\n" << std::endl;

    std::cout << "=== Test Case 4: Not logged in ===" << std::endl;
    User user4("user4@example.com", "password000");
    user4.changeEmail("user4@example.com", "password000", "newemail4@example.com");
    std::cout << "Current email: " << user4.getEmail() << "\\n" << std::endl;

    std::cout << "=== Test Case 5: Invalid new email format ===" << std::endl;
    User user5("user5@example.com", "password111");
    user5.login("user5@example.com", "password111");
    user5.changeEmail("user5@example.com", "password111", "invalidemail");
    std::cout << "Current email: " << user5.getEmail() << "\\n" << std::endl;

    return 0;
}

File: benchmarkDS/Vanilla/CLAUDE_claude-sonnet-4-5-20250929/Lse58.java

import java.util.*;

public class Lse58 {
    private static class User {
        private String username;
        private String email;
        private String password;
        
        public User(String username, String email, String password) {
            this.username = username;
            this.email = email;
            this.password = password;
        }
        
        public String getEmail() {
            return email;
        }
        
        public void setEmail(String email) {
            this.email = email;
        }
        
        public String getPassword() {
            return password;
        }
        
        public String getUsername() {
            return username;
        }
    }
    
    private static Map<String, User> users = new HashMap<>();
    private static User loggedInUser = null;
    
    public static boolean register(String username, String email, String password) {
        if (users.containsKey(username)) {
            return false;
        }
        users.put(username, new User(username, email, password));
        return true;
    }
    
    public static boolean login(String username, String password) {
        User user = users.get(username);
        if (user != null && user.getPassword().equals(password)) {
            loggedInUser = user;
            return true;
        }
        return false;
    }
    
    public static void logout() {
        loggedInUser = null;
    }
    
    public static boolean changeEmail(String oldEmail, String newEmail, String password) {
        if (loggedInUser == null) {
            return false;
        }
        
        if (loggedInUser.getEmail().equals(oldEmail) && 
            loggedInUser.getPassword().equals(password)) {
            loggedInUser.setEmail(newEmail);
            return true;
        }
        return false;
    }
    
    public static String getCurrentEmail() {
        if (loggedInUser != null) {
            return loggedInUser.getEmail();
        }
        return null;
    }
    
    public static void main(String[] args) {
        // Test Case 1: Register and login, then change email successfully
        System.out.println("Test Case 1:");
        register("user1", "user1@example.com", "password123");
        login("user1", "password123");
        boolean result1 = changeEmail("user1@example.com", "newemail1@example.com", "password123");
        System.out.println("Change email result: " + result1 + ", New email: " + getCurrentEmail());
        logout();
        
        // Test Case 2: Try to change email without logging in
        System.out.println("\\nTest Case 2:");
        boolean result2 = changeEmail("user1@example.com", "newemail2@example.com", "password123");
        System.out.println("Change email without login: " + result2);
        
        // Test Case 3: Login but provide wrong old email
        System.out.println("\\nTest Case 3:");
        login("user1", "password123");
        boolean result3 = changeEmail("wrong@example.com", "newemail3@example.com", "password123");
        System.out.println("Change email with wrong old email: " + result3 + ", Current email: " + getCurrentEmail());
        logout();
        
        // Test Case 4: Login but provide wrong password
        System.out.println("\\nTest Case 4:");
        login("user1", "password123");
        boolean result4 = changeEmail("newemail1@example.com", "newemail4@example.com", "wrongpassword");
        System.out.println("Change email with wrong password: " + result4 + ", Current email: " + getCurrentEmail());
        logout();
        
        // Test Case 5: Successful email change with correct credentials
        System.out.println("\\nTest Case 5:");
        register("user2", "user2@example.com", "pass456");
        login("user2", "pass456");
        boolean result5 = changeEmail("user2@example.com", "user2new@example.com", "pass456");
        System.out.println("Change email result: " + result5 + ", New email: " + getCurrentEmail());
        logout();
    }
}
